src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -9276,6 +9287,114 @@ def quantiles_along_axis0(
+        | index |    a |    b |
+        +-------+------+------+                    +-------+------+------+
+        |  0.25 | 0.25 | 1.25 |                    | index |    a |    b |
+        +-------+------+------+                    +-------+------+------+
+                                 -- UNION ALL -->  |  0.25 | 0.25 | 1.25 |
+        +-------+------+------+                    +-------+------+------+
+        | index |    a |    b |                    |  0.75 | 0.75 | 1.75 |
+        +-------+------+------+                    +-------+------+------+
+        |  0.75 | 0.75 | 1.75 |
+        +-------+------+------+
+
+        When there is a large number of q (as is the case for most uses of qcut), the number of
+        UNION operations increases dramatically, and may cause Snowpark to create temporary tables.
+        This greatly increases latency.
+
+        When the input frame has a single column, we can eliminate UNION ALL operations
+        by producing a single row where the columns are the different quantiles. Since there is
+        only a single row, we can do a relatively cheap UNPIVOT to make the result a single column.
+
+        >>> pd.Series([0, 1], name="b").quantile([0.25, 0.75])
+        +------+------+                                              +-------+------+
+        |   q1 |   q2 |                                              | index |    b |
+        +------+------+  -- UNPIVOT(b FOR quantile IN (q1, q2)) -->  +-------+------+
+        | 1.25 | 1.75 |                                              |    q1 | 1.25 |
+        +------+------+                                              +-------+------+
+                                                                     |    q2 | 1.75 |
+                                                                     +-------+------+
+
+        qcut() can drop the index column afterwards, but quantile() keeps it.
+        """
+        if index is not None:
+            # Snowpark UNPIVOT requires these to be strings
+            index = list(map(str, index))
+        original_frame = self._modin_frame
+        col_label = original_frame.data_column_pandas_labels[0]
+        col_ident = original_frame.data_column_snowflake_quoted_identifiers[0]
+        new_labels = [str(quantile) for quantile in q]
+        new_idents = (
+            original_frame.ordered_dataframe.generate_snowflake_quoted_identifiers(
+                pandas_labels=new_labels
+            )
+        )
+        ordered_dataframe = original_frame.ordered_dataframe.agg(
+            *[
+                column_quantile(col(col_ident), interpolation, quantile).as_(new_ident)
+                for new_ident, quantile in zip(new_idents, q)
+            ]
+        )
+        index_ident = ordered_dataframe.generate_snowflake_quoted_identifiers(
+            pandas_labels=[None]
+        )[0]
+        # In order to preserve index labels, we call unpivot directly instead of using transpose_single_row
```

Comment:
is this a bug in transpose_single_row? Should we fix it?

Reply:
What I mean by this is that there's nowhere in `transpose_single_row` that would easily let us set the labels of the resulting index (the names of the columns before UNPIVOT) to be our desired values. This matters in cases like `df.describe`, where the quantiles have labels like "25%" and "50%" rather than the usual decimal "0.25", "0.5", etc. We would have to do a JOIN later to add a column with the correct index. Calling unpivot directly also lets us avoid the JSON serialization/deserialization.

I may end up using `transpose_single_row` anyway, since I'm having difficulty getting `UNPIVOT` to preserve the order of the passed in quantiles in the result column. This would force us to do a JOIN anyway to ensure the index is respected.